src/mame/barcrest/mpu1.cpp

```diff
@@ -35,85 +53,233 @@
 		m_pia1(*this, "pia1"),
 		m_pia2(*this, "pia2"),
 		m_lamps(*this, "lamp%u", 0U),
-		m_reels(*this, "emreel%u", 1U),
 		m_dac(*this, "dac"),
-		m_samples(*this, "samples")
+		m_samples(*this, "samples"),
+		m_pia3(*this, "pia3"),
+		m_pia4(*this, "pia4"),
+		m_nvram(*this, "nvram", 0x100, ENDIANNESS_BIG)
 	{ }
 
-	DECLARE_INPUT_CHANGED_MEMBER(coin_input);
+	void mpu12_base(machine_config &config);
 
-	void mpu1(machine_config &config);
-	void mpu1_lg(machine_config &config);
+	DECLARE_INPUT_CHANGED_MEMBER(coin_input);
 
 protected:
+	enum class lamp_flags : uint8_t
+	{
+		IGNORE_RELAY = 0,
+		RELAY_OFF = 0x1,
+		RELAY_ON = 0x2,
+		RELAY_BOTH = 0x3
+	};
+
 	virtual void machine_start() override;
 	virtual void machine_reset() override;
 
-private:
-	enum { STEPS_PER_SYMBOL = 20 };
-
-	template <unsigned Lamp> void pia_lamp_w(int state) { m_lamps[Lamp] = state; }
-	template <unsigned Reel> void reel_sample_cb(uint8_t state);
-
 	TIMER_DEVICE_CALLBACK_MEMBER(nmi);
 	TIMER_CALLBACK_MEMBER(change_pia2a_bit7);
 
-	void pia1_portb_w(uint8_t data);
-	void pia1_portb_lg_w(uint8_t data);
-	uint8_t pia2_porta_r();
-	void pia2_porta_w(uint8_t data);
-	void pia2_portb_w(uint8_t data);
-	uint8_t reel_pos_r(uint8_t reel);
+	void pia_lamp_update(int data) { update_pia_lamps(); }
+	uint8_t nvram_r(offs_t offset) { return m_nvram[offset] & 0xf; }
+	void nvram_w(offs_t offset, uint8_t data) { m_nvram[offset] = data & 0xf; }
+	virtual void update_pia_lamps() {}
+	void set_lamp(bool state, int lamp1, int lamp2 = 0, lamp_flags flags = lamp_flags::IGNORE_RELAY);
+	void update_lamp_relay(uint8_t state);
 	void payout_cash_w(bool state);
 	void payout_token_w(bool state);
 	void meter_w(int meter, bool state);
+	void meter_tick_w(int meter, bool state);
 	void coin_lockout_w(bool state);
 
 	void mpu1_map(address_map &map);
-
-	uint8_t m_reel_select;
+	void mpu2_map(address_map &map);
+	void mpu2_nvram_map(address_map &map);
+
+	uint8_t m_lamp_relay;
+	lamp_flags m_lamp_flags_pia2b[8];
+	lamp_flags m_lamp_flags_pia3b[7];
+	lamp_flags m_lamp_flags_pia4b[8];
+	lamp_flags m_lamp_flags_pia1cb2;
+	lamp_flags m_lamp_flags_pia2ca2;
+	lamp_flags m_lamp_flags_pia2cb2;
+	lamp_flags m_lamp_flags_pia3ca2;
+	lamp_flags m_lamp_flags_pia3cb2;
+	lamp_flags m_lamp_flags_pia4ca2;
+	lamp_flags m_lamp_flags_pia4cb2;
 	bool m_pia2a_select;
-	bool m_prev_payout[2];
+	bool m_payout_state[2];
 	uint8_t m_pia2a_bit7_value;
 	emu_timer *m_change_pia2a_bit7_timer;
 
 	required_device<cpu_device> m_maincpu;
 	required_device<timer_device> m_nmi_timer;
 	required_device<pia6821_device> m_pia1;
 	required_device<pia6821_device> m_pia2;
-	output_finder<13> m_lamps;
-	required_device_array<em_reel_device, 4> m_reels;
+	output_finder<56> m_lamps;
 	required_device<dac_1bit_device> m_dac;
 	required_device<fruit_samples_device> m_samples;
+
+	// MPU2 devices
+	optional_device<pia6821_device> m_pia3;
+	optional_device<pia6821_device> m_pia4;
+	memory_share_creator<uint8_t> m_nvram;
```

Comment:
If these are for the MPU2, can you move them out of the base class and make them `required_device` rather than `optional_device`?  It’s better to avoid `optional_device` if it isn’t really necessary.

Reply:
Fixed.